### ISSUE TITLE
[Feature Request] Optimizer should not discard data before `revlog_start_date`

### ISSUE THREAD

**Issue by nb9618** [open-spaced-repetition/fsrs4anki#320]
**Which module is related to your feature request?**
Optimizer

**Is your feature request related to a problem? Please describe.**
Using `revlog_start_date` discards review log data and makes it impossible for the optimizer to correctly calculate memory states for affected cards. This leads to increase of log loss and probably to worse optimization result. However, I think that the ability to optimize parameters using recent review data instead of the entire review history is still useful for the following reasons:

-  User's review habits may change. This includes grading with 2 vs 4 buttons, setting different learning intervals, restricting the time to recall the answer before failing the card, etc.
- For some study subjects, the understanding of the subject in its entirety increases together with the number of individual known facts. As the overall picture becomes more clear, individual cards may become easier. This would affect calculation of difficulty as well as time spent per card.

**Describe the solution you'd like**
Review data should not be discarded. Instead, only those reviews that happened after `revlog_start_date` should contribute to calculation of log loss during training. Additionally, a new option, something like `evaluation_start_date`, should be introduced for the purpose of computing the metrics after training is finished. Using the combination of these two options, users would be able to optimize for the entire review log and for a chosen time frame, and compare the performance of both optimizations over the same time period.

**Describe alternatives you've considered**
If someone's reviewing behavior has changed, it does not mean that the review data before that change becomes completely useless. For this reason, instead of excluding old reviews from log loss calculation, weights could be assigned to reviews based on how old they are. However, I'm not sure which function should be used for this purpose. The best I can come up with is, supposing that user has set a one month interval, to set weight to 1 for the first month, then to 0.5 for next two months, then 0.25 for next four months and so on.

**Additional context**
Sadly, I'm not skillful enough to implement this myself.

**Comment by user1823** [comment]
> Review data should not be discarded. Instead, only those reviews that happened after `revlog_start_date` should contribute to calculation of log loss during training. Additionally, a new option, something like `evaluation_start_date`, should be introduced for the purpose of computing the metrics after training is finished. Using the combination of these two options, users would be able to optimize for the entire review log and for a chosen time frame, and compare the performance of both optimizations over the same time period.

I agree with the main idea. I am also of the opinion that the evaluation data should not be necessarily the same as the training data.

My idea is that the training data should consist of only those reviews which can be efficiently dealt with by the algorithm. However, the evaluation data should contain ALL the data that affects future reviewing of any card.

For example, in my opinion,
- All the ratings of a card before a manual reset should be discarded because they are likely to be outliers (and thus, not fit for inclusion in the training data) and they don't affect the future reviewing (and thus, not fit for inclusion in the evaluation data).
- All the ratings of a card before the `revlog_start_date` should be removed from the training data but they should be included in the evaluation data (because they still affect the future reviewing).
- The outliers discussed in [this issue](https://github.com/open-spaced-repetition/fsrs4anki/issues/275) should be removed from the training data but they should be included in the evaluation data (because they still affect the future reviewing).

By the way, because I think that ALL the data that affects future reviewing of any card should be included in the evaluation data, I don't think that we need a new variable such as `evaluation_start_date`.

> If someone's reviewing behavior has changed, it does not mean that the review data before that change becomes completely useless. For this reason, instead of excluding old reviews from log loss calculation, weights could be assigned to reviews based on how old they are.

This idea makes sense. But, I don't have any good idea to implement it.

**Comment by nb9618** [comment]
> By the way, because I think that ALL the data that affects future reviewing of any card should be included in the evaluation data, I don't think that we need a new variable such as evaluation_start_date.
The reason I want this option is because it will allow to compare performance of different optimizations on recent reviews. While optimization over entire history may produce better fitting for old reviews and better overall metrics, it doesn't necessary mean that it will be better for recent reviews, compared to optimization limited by `revlog_start_date`. Performance of recent reviews is important because it better represents what would happen when the user will do new reviews, that is the ability of algorithm to make predictions for new data it hasn't been trained on.

**Comment by user1823** [comment]
> > By the way, because I think that ALL the data that affects future reviewing of any card should be included in the evaluation data, I don't think that we need a new variable such as evaluation_start_date.
> 
> The reason I want this option is because it will allow to compare performance of different optimizations on recent reviews.

It makes sense.

I would like to add one point: This `evaluation_start_date` variable should be added just before the evaluation begins (i.e., in section 4) so that we can change it to produce evaluations for different sets of data without needing to train the parameters again.

Also, as I said earlier, the evaluation data should be different from the training data. So, this `evaluation_start_date` should be used to filter the evaluation data and not the training data.

**Comment by L-M-Sherlock** [comment]
> Review data should not be discarded. Instead, only those reviews that happened after should contribute to calculation of log loss during training.

I don't know how to contribute the loss during training with reviews only after the revlog_start_date, without discarding review data from before this date.

**Comment by user1823** [comment]
> > Review data should not be discarded. Instead, only those reviews that happened after should contribute to calculation of log loss during training.
> 
> I don't know how to contribute the loss during training with reviews only after the revlog_start_date, without discarding review data from before this date.

@nb9618 means to say that the optimizer should maintain two datasets. The dataset used for training should discard the review data before the `revlog_start_date` but the dataset used for evaluation should not discard the review data.

**Comment by L-M-Sherlock** [comment]
> Using `revlog_start_date` discards review log data and makes it impossible for the optimizer to correctly calculate memory states for affected cards

Another solution is only keeping the cards whose first reviews happen after `revlog_start_date`.

**Comment by user1823** [comment]
> > Using `revlog_start_date` discards review log data and makes it impossible for the optimizer to correctly calculate memory states for affected cards
> 
> Another solution is only keeping the cards whose first reviews happen after `revlog_start_date`.

I think that this happens with the current version of the optimizer also. This is because it filters out the reviews in which the first rating is not of learn type.
Another supporting fact is [this issue](https://github.com/open-spaced-repetition/fsrs4anki/issues/277).

By the way, I think that the main concern of @nb9618 is about the results in the evaluation section.

**Comment by nb9618** [comment]
> I don't know how to contribute the loss during training with reviews only after the revlog_start_date, without discarding review data from before this date.

I was thinking that `weight` parameter of `nn.BCELoss` could be used. However, I know virtually nothing about machine learning and PyTorch in particular so I may be completely wrong.

>Another solution is only keeping the cards whose first reviews happen after revlog_start_date.

In collections with many old cards and not many new ones this will filter out the majority of cards. Also, this will mean that optimizer will see no reviews with intervals longer than time passed since `revlog_start_date`, and its performance for reviews with high stabilities will be poor.

>By the way, I think that the main concern of @nb9618 is about the results in the evaluation section.

I'm very concerned with the fact that using `revlog_start_date` always increases log loss, despite that reducing the number of potentially contradicting data points should instead decrease it. This is an indication of that the optimizer is presented with data which it is not designed to efficiently process and therefore the optimization result may be incorrect.

**Comment by L-M-Sherlock** [comment]
> I was thinking that `weight` parameter of `nn.BCELoss` could be used. However, I know virtually nothing about machine learning and PyTorch in particular so I may be completely wrong.

It is possible to mask these reviews based on their review data. But I need to add this field to the trainset.

> Also, this will mean that optimizer will see no reviews with intervals longer than time passed since `revlog_start_date`, and its performance for reviews with high stabilities will be poor.

Using the old reviews with high stability will make the performance worse.

> I'm very concerned with the fact that using `revlog_start_date` always increases log loss, despite that reducing the number of potentially contradicting data points should instead decrease it. This is an indication of that the optimizer is presented with data which it is not designed to efficiently process and therefore the optimization result may be incorrect.

The log loss is affected by the distribution of data. It is very common.

**Comment by user1823** [comment]
Well, @nb9618, your original post contains two suggestions: one is related to the evaluation section and one is concerned with using even the older review data for training (with appropriate weights).

So, I think that we should split these into two separate issues.

The first part (related to evaluation) is easier to implement and we have discussed it extensively. So, I think that this issue should be used for discussing that.

For discussion about using review data older than revlog_start_date during training (with appropriate weights), I recommend creating a new issue. The main reason being that none of us currently have a good idea to implement it.

**Comment by user1823** [comment]
I am saving this discussion here so that it doesn't get forgotten after https://github.com/open-spaced-repetition/fsrs4anki/issues/270 is closed.

> All the ratings of suspended cards should be filtered out (because they are more likely to be outliers).

They are not outliers in my case. I archived some cards via `suspend` in my collection. The review logs of these cards are useful.

_Originally posted by @L-M-Sherlock in https://github.com/open-spaced-repetition/fsrs4anki/issues/270#issuecomment-1612640534_

Now, I think that we should leave it to the user whether they want to train the parameters on the suspended cards or not. They can easily control this by exporting either all cards or only unsuspended cards from the Anki Browser.

_Originally posted by @user1823 in https://github.com/open-spaced-repetition/fsrs4anki/issues/270#issuecomment-1612689617_

We might not be filtering out the suspended cards in the training data. But, when we will later use a separate dataset for evaluation, we should **filter out the suspended cards in the evaluation data**. The reason being that even if the revlogs of the suspended cards can be helpful in training the parameters, those revlogs are of no value in the evaluation process (because those cards are not being reviewed).

_Originally posted by @user1823 in https://github.com/open-spaced-repetition/fsrs4anki/issues/270#issuecomment-1612748258_

**Comment by L-M-Sherlock** [comment]
In #334, I have decided to discard all review data that contain reviews happened before `revlog_start_date`. So this issue will be closed.

But, I think this suggestion is rational:

> If someone's reviewing behavior has changed, it does not mean that the review data before that change becomes completely useless. For this reason, instead of excluding old reviews from log loss calculation, weights could be assigned to reviews based on how old they are. However, I'm not sure which function should be used for this purpose. The best I can come up with is, supposing that user has set a one month interval, to set weight to 1 for the first month, then to 0.5 for next two months, then 0.25 for next four months and so on.

It is technically feasible to weight the loss depending on the latency of reviews. It would be an exponentially decay. But it would also make the issue in #187 worse.

> Now, I think that we should leave it to the user whether they want to train the parameters on the suspended cards or not. They can easily control this by exporting either all cards or only unsuspended cards from the Anki Browser.

I will open a new issue to track it.
